src/rpi.py

```diff
@@ -0,0 +1,132 @@
+    cmd_result = subprocess.run(cmd, stdout=subprocess.PIPE, check = False)
+    if cmd_result.returncode != 0:
+        raise RuntimeError("Failed to generate RPI key pair")
+
+    fingerprint = hashlib.md5(cmd_result.stdout.split()[1]).hexdigest()
+    shutil.move(tmpdir, RPI_DIR + os.sep + fingerprint)
+    response = {"fingerprint": fingerprint}
+
+    LOG.info("%s successfully generated a new RPI key %s", user, fingerprint)
+
+    return response
+
+
+def key(key_id):
+    for entry in os.listdir(RPI_DIR):
+        if entry == key_id:
+            with open(os.path.join(RPI_DIR, entry, 'public.pem'), 'r',
+                      encoding='utf-8') as pubkey:
+                response = {
+                    "key": pubkey.read(),
+                }
+                return response
+
+    LOG.info("Failed to find key %s", key_id)
+    return {"error": f"RPI key {key_id} is unknown"}, 404
+
+def sha256sum(payload):
+    sha256_hash = hashlib.sha256()
+    with open(payload,"rb") as f:
+        for byte_block in iter(lambda: f.read(4096),b""):
+            sha256_hash.update(byte_block)
+        return sha256_hash.hexdigest()
+
+# Based on https://github.com/raspberrypi/rpi-eeprom/blob/master/rpi-eeprom-digest
+
+# sha256sum "${IMAGE}" | awk '{print $1}' > "${OUTPUT}"
+# echo "ts: $(date -u +%s)" >> "${OUTPUT}"
+# "${OPENSSL}" dgst -sign "${KEY}" -keyform PEM -sha256 -out "${SIG_TMP}" "${IMAGE}"
+# echo "rsa2048: $(xxd -c 4096 -p < "${SIG_TMP}")" >> "${OUTPUT}"
+
+def sign(body, user):
+    key_id = body["key_id"]
+
+    for entry in os.listdir(RPI_DIR):
+        if entry == key_id:
+            try:
+                payload = binascii.a2b_base64(body["payload"])
+            except Exception as ex:
+                return {"error": f"Failed to base64-decode payload: {ex}"}, 400
+
+            tmpinput = tempfile.mkstemp(prefix="rpi-")[1]
+            with open(tmpinput, 'wb') as file:
+                file.write(payload)
+                tmpfile = tempfile.mkstemp(prefix="rpi-")[1]
+                cmd = [ "openssl", "dgst", "-sign",
+                       os.path.join(RPI_DIR, entry, "private.pem"),
+                       "-keyform", "PEM",
+                       "-sha256", "-out", tmpfile, tmpinput ]
+                cmd_result = subprocess.run(cmd, check = False)
+                if cmd_result.returncode != 0:
+                    raise RuntimeError("Failed to sign payload")
+                tmpout = tempfile.mkstemp(prefix="rpi-")[1]
+                with open(tmpout, 'a', encoding='utf-8') as output:
+                    output.write(sha256sum(tmpinput))
+                    output.write("\n")
+                    output.write("ts: " +
+                                 str(int(datetime.utcnow().timestamp())))
+                    output.write("\n")
+                    hex_string=""
+                    with open(tmpfile, 'rb') as ifile:
+                        while True:
```

Comment:
Similar as above, do we expect the file to be big? I do not think it would be a problem to just read and hexlify it in memory in one go, just capping the max size read from `ifile`.

Reply:
It would also work, but why not just remove the file size limitation now?